### PR TITLE
GH-43396: [Java] Remove/replace jsr305

### PR DESCRIPTION
**Pull request by laurentgo**
### Rationale for this change

jsr305 is not maintained anymore and is unlikely to support JPMS. The classes are also in the javax. namespace which is known to cause issues as well.

### What changes are included in this PR?

Replace most uses of jsr305 with the equivalent checker framework annotations, and remove usage of `@ThreadSafe.`

### Are these changes tested?

CI/CD

### Are there any user-facing changes?

None

* GitHub Issue: #43396

### CODE REVIEW

**Comment by laurentgo** [comment]
@github-actions crossbow submit -g java

**Review by danepitkin — APPROVED** [review]
LGTM! I'll wait for CI to pass before merging.

**Review by lidavidm — APPROVED** [review]
I thought Protobuf generated code depended on this at some point but apparently not anymore.

**Review by vibhatha — APPROVED** [review]

**Comment by danepitkin** [comment]
> I thought Protobuf generated code depended on this at some point but apparently not anymore.

Turns out this line omits the annotation that protobuf was adding: https://github.com/apache/arrow/pull/43139/files#diff-d8225ebcfc11b480a6e4f54e183b67c3ead51635a167c106d928c2abf1f9ef66R492

**Pull request merged**
